utils/controller_debug.py

```diff
@@ -0,0 +1,1058 @@
+# command or response.  This function removes those characters
+# before returning.
+# See debug.cpp in the controller source for more detail on
+# command framing.
+def GetResp(DbgPrint):
+    dat = []
+    esc = False
+    DbgPrint("Getting resp: ", end="")
+    while True:
+        x = ser.read(1)
+        if len(x) < 1:
+            DbgPrint("timeout")
+            return dat
+        x = ord(x)
+        DbgPrint("0x%02x" % x, end=" ")
+
+        if esc:
+            esc = False
+            dat.append(x)
+            continue
+
+        if x == ESC:
+            esc = True
+            continue
+
+        if x == TERM:
+            DbgPrint()
+            return dat
+        dat.append(x)
+
+
+# This formats a binary command and sends it to the system.
+# It then waits for and returns a response.
+#
+#  op - The command code to send.  See the list at the top
+#       of the file.
+
+#  data - Zero or more bytes of data to be sent with the command
+
+#  timeout - How long (seconds) to wait for the response.  If
+#            not specified then a reasonable system default is used
+def SendCmd(op, data=[], timeout=None):
+    global showSerial, ser
+
+    if showSerial:
+        DbgPrint = print
+    else:
+
+        def DbgPrint(S="", end="\n"):
+            pass
+
+    buff = [op] + data
+
+    crc = crc16.calc(buff)
+    buff += Split16(crc)
+
+    S = "CMD: "
+    for b in buff:
+        S += "0x%02x " % b
+    DbgPrint(S)
+
+    buff = EscCmd(buff)
+    S = "ESC: "
+    for b in buff:
+        S += "0x%02x " % b
+    DbgPrint(S)
+
+    ser.write(bytearray(buff))
+    if timeout != None:
+        DbgPrint("Setting timeout to %.1f" % timeout)
+        oldto = ser.timeout
+        ser.timeout = timeout
+
+    rsp = GetResp(DbgPrint)
+    if timeout != None:
+        ser.timeout = oldto
+
+    if len(rsp) < 3:
+        raise Error("Invalid response, too short")
+
+    crc = crc16.calc(rsp[:-2])
+    rcrc = Build16(rsp[-2:])[0]
+    if crc != rcrc:
+        raise Error(
+            "CRC error on response, calculated 0x%04x received 0x%04x" % (crc, rcrc)
+        )
+
+    if rsp[0]:
+        raise Error("Error %d (0x%02x)" % (rsp[0], rsp[0]))
+        return []
+    return rsp[1:-2]
+
+
+def ReSync():
+    cmd = [TERM, TERM]
+    ser.write(bytearray(cmd))
+    time.sleep(0.1)
+    ser.reset_input_buffer()
+
+
+def I2F(ival):
+    s = struct.pack("I", ival & 0xFFFFFFFF)
+    return struct.unpack("f", s)[0]
+
+
+def F2I(fval):
+    s = struct.pack("f", fval)
+    return struct.unpack("I", s)[0]
+
+
+# Takes as input a list of integer values and
+# splits them into a list of bytes which is returned.
+#
+# val - The input list of ints
+# len - Length (in bytes) of the input ints.
+# le  - Little endian format if true
+def SplitInt(val, len=4, le=True):
+    ret = []
+
+    if isinstance(val, list):
+        for v in val:
+            ret += SplitInt(v, len, le)
+    else:
+        for i in range(len):
```

Comment:
Looks like this can be simply "ret = val.to_bytes(len, byteorder='little' if le else 'big')"

Reply:
Oh, that's nice.  I have mostly been working in Python 2.7 and wasn't aware of that function.  That's new to 3.x I guess.

Reply:
Hum, to_bytes only sees to work in individual ints, not lists.  Not sure if it really helps.  I'll experiment with it some, but may stick with the current method if it ends up being more wordy doing it the other way